forch/__main__.py

```diff
@@ -0,0 +1,106 @@
+    # TODO: 1) use protobuf after entire forch config is converted
+    #       2) clean up places where individual forch config sections are converted by dict_proto
+    #          instead of direct access from forch config proto obj
+    config_root = os.getenv('FORCH_CONFIG_DIR', '.')
+    config_path = os.path.join(config_root, _FORCH_CONFIG_DEFAULT)
+    LOGGER.info('Reading config file %s', os.path.abspath(config_path))
+    try:
+        with open(config_path, 'r') as stream:
+            return yaml.safe_load(stream)
+    except Exception as e:
+        LOGGER.error('Cannot load config: %s', e)
+        return None
+
+
+def show_error(error, path, params):
+    """Display errors"""
+    return f"Cannot initialize forch: {str(error)}"
+
+
+def run_forchestrator():
+    """main function to start forch"""
+    configure_logging()
+
+    config = load_config()
+    if not config:
+        LOGGER.error('Invalid config, exiting.')
+        sys.exit(1)
+
+    forchestrator = Forchestrator(config)
+    http_server = forch.http_server.HttpServer(
+        config.get('http', {}), forchestrator.get_local_port())
+
+    try:
+        forchestrator.initialize()
+        http_server.map_request('system_state', forchestrator.get_system_state)
+        http_server.map_request('dataplane_state', forchestrator.get_dataplane_state)
+        http_server.map_request('switch_state', forchestrator.get_switch_state)
+        http_server.map_request('cpn_state', forchestrator.get_cpn_state)
+        http_server.map_request('process_state', forchestrator.get_process_state)
+        http_server.map_request('host_path', forchestrator.get_host_path)
+        http_server.map_request('list_hosts', forchestrator.get_list_hosts)
+        http_server.map_request('sys_config', forchestrator.get_sys_config)
+        http_server.map_request('', http_server.static_file(''))
+    except Exception as e:
+        LOGGER.error("Cannot initialize forch: %s", e, exc_info=True)
+        http_server.map_request('', functools.partial(show_error, e))
+    finally:
+        http_server.start_server()
+
+    if forchestrator.initialized():
+        forchestrator.main_loop()
+    else:
+        try:
+            http_server.join_thread()
+        except KeyboardInterrupt:
+            LOGGER.info('Keyboard interrupt. Exiting.')
+
+    LOGGER.warning('Exiting program')
+    http_server.stop_server()
+    forchestrator.stop()
+
+
+def parse_args(raw_args):
+    """Parse command line arguments"""
+    parser = argparse.ArgumentParser(prog='forch', description='Process some integers.')
+    parser.add_argument('-V', '--version', action='store_true', help='print version and exit')
+    parsed = parser.parse_args(raw_args)
+    return parsed
+
+
+def main():
+    """Main program"""
+    args = parse_args(sys.argv[1:])
+
+    if args.version:
+        print(f'Forch {__version__}')
+        sys.exit()
+
+    run_forchestrator()
+
+
+if __name__ == '__main__':
```

Comment:
this seems redundant -- why define main() at all?

Reply:
bin/forch runs forch as a module so there is __main__ here.

'main()' will be run in production environment. In setup.cfg, an entry point of __main__.py:main() is added. When installing the package, a python script in bin directory will be generated, in which __main__.py:main() will be called. For the entry point settings in setup.cfg, we can only specify a function instead of a file.